pkg/install/install.go

```diff
@@ -0,0 +1,464 @@
+		logrus.Info("Installing any TLS assets")
+		mkdir("/host/etc/cni/net.d/calico-tls")
+		if err := copyFile(fmt.Sprintf("%s/%s", c.TLSAssetsDir, "etcd-ca"), "/host/etc/cni/net.d/calico-tls/etcd-ca"); err != nil {
+			logrus.Warnf("Missing etcd-ca")
+		}
+		if err := copyFile(fmt.Sprintf("%s/%s", c.TLSAssetsDir, "etcd-cert"), "/host/etc/cni/net.d/calico-tls/etcd-cert"); err != nil {
+			logrus.Warnf("Missing etcd-cert")
+		}
+		if err := copyFile(fmt.Sprintf("%s/%s", c.TLSAssetsDir, "etcd-key"), "/host/etc/cni/net.d/calico-tls/etcd-key"); err != nil {
+			logrus.Warnf("Missing etcd-key")
+		}
+	}
+
+	// Place the new binaries if the directory is writeable.
+	dirs := []string{"/host/opt/cni/bin", "/host/secondary-bin-dir"}
+	for _, d := range dirs {
+		if unix.Access(d, unix.W_OK) != nil {
+			logrus.Infof("%s is not writeable, skipping", d)
+			continue
+		}
+
+		// Iterate through each binary we might want to install.
+		// TODO: Include all the binaries we ship.
+		binaries := []string{"calico", "calico-ipam", "bandwidth", "flannel", "loopback", "host-local", "portmap", "tuning"}
+		for _, binary := range binaries {
+			target := fmt.Sprintf("%s/%s", d, binary)
+			source := fmt.Sprintf("/opt/cni/bin/%s", binary)
+			if strings.Contains(binary, "calico") {
+				// For Calico binaries, we copy over the install binary. It includes the code
+				// for each, and the name of the binary determined which is executed.
+				source = "/opt/cni/bin/install"
+			}
+			if c.skipBinary(binary) {
+				continue
+			}
+			if fileExists(target) && c.UpdateCNIBinaries == "true" {
+				logrus.Infof("Skipping installation of %s", target)
+				continue
+			}
+			if err := copyFile(source, target); err != nil {
+				logrus.WithError(err).Errorf("Failed to install %s", target)
+				os.Exit(1)
+			}
+			logrus.Infof("Installed %s", target)
+		}
+
+		logrus.Infof("Wrote Calico CNI binaries to %s\n", d)
+
+		// Print CNI plugin version
+		cmd := exec.Command(d+"/calico", "-v")
+		var out bytes.Buffer
+		cmd.Stdout = &out
+		err := cmd.Run()
+		if err != nil {
+			logrus.WithError(err).Warnf("Failed getting CNI plugin version")
+		}
+		logrus.Infof("CNI plugin version: %s", out.String())
+	}
+
+	if kubecfg != nil {
+		// If running as a Kubernetes pod, then write out a kubeconfig for the
+		// CNI plugin to use.
+		writeKubeconfig(kubecfg)
+	}
+
+	// Write a CNI config file.
+	writeCNIConfig(c)
+
+	// Unless told otherwise, sleep forever.
+	// This prevents Kubernetes from restarting the pod repeatedly.
+	shouldSleep := getEnv("SLEEP", "true")
+	logrus.Infof("Done configuring CNI.  Sleep= %v", shouldSleep)
+	for shouldSleep == "true" {
+		// Kubernetes Secrets can be updated.  If so, we need to install the updated
+		// version to the host. Just check the timestamp on the certificate to see if it
+		// has been updated.  A bit hokey, but likely good enough.
+		filename := c.TLSAssetsDir + "/etcd-cert"
+		info, err := os.Stat(filename)
+		if err == nil {
```

Comment:
Should we be handling specific errors (namely not exists?)

Reply:
Changed to if file does not exist then sleep, else do logic